### DISCUSSION TITLE
Configuring browser on local

### DISCUSSION THREAD

**Discussion by carabusu** [eliasnogueira/selenium-java-lean-test-architecture#77] [Q&A] [answered]
I changed target=local (in general.properties) and browser = firefox (in local.properties) but the browser that will start is chrome. I do not have any other env variables with browser name.
How can I do so that I can use other browser than chrome on local ?

**Comment by eliasnogueira** [comment, marked as answer]
Hi @carabusu ,
Thanks for your issue.

The chrome is in use because it's the default browser set in the [BaseWeb](https://github.com/eliasnogueira/selenium-java-lean-test-architecture/blob/master/src/test/java/com/eliasnogueira/BaseWeb.java#L50) class.
The single test execution via the test class was meant to execute via `chrome` all the time.

When you want to run in different browsers we must either change the `@Optional` parameter to `firefox` or run the tests via a TestNG test suite setting the parameter value as Firefox
```
<parameter name="browser" value="firefox"/>
```

You can easily change the existing test suite [local.xml](https://github.com/eliasnogueira/selenium-java-lean-test-architecture/blob/master/src/test/resources/suites/local.xml).

Please, let me know if you have any other questions.

**Comment by vnuta9** [comment]
Hi @eliasnogueira 
It looks like POM is being implemented. For our purpose, it is little unnecessary and also over kill. How can I use the f/w with out POM, just simply script files with one or two utility classes ?
I am "believing" that there are no help implementing any fluent waits on the Web Element actions since I am working on a eCommerce website made of node.js technology.
And from @carabusu question and the answer he got, I am assuming that by changing the testNG.xml suite browser parameter value to 'chrome' to 'firefox' and vice versa to run on these or any other browsers like 'edge'. I guess NOT but do we have the ability to run parallel running of two distinct browsers on a single windows platform by configuring the same testNG.xml file. Do we ?

Appreciate your efforts in producing the f/w and for your response to my query.

--Venkata Nutalapati

**Reply by eliasnogueira** [reply]
Hi @vnuta9,

Thank you for your feedback!

### About the POM approach
The POM would never be necessary. You might think that you would spend more time creating the POM structure (and you will), but later you will miss it as your application evolves and you need to maintain it.

I have never been a fan of utility classes that directly encapsulate the command Selenium actions like click, sendKeys, etc...

If you want to use directly the driver instance you can use `DriverMagager().getDriver()` and you will be able to use any Selenium commands, like

```java
WebDriver driver = DriverMagager().getDriver();
driver.findElements(By.id).sendKeys("user1234);
// and so son
```

### About the fluent waits

It is worth creating a strategy for it, as normally e-commerce have asynchronous requests. It can be using the Fluent Waits or the Explicitly Waits.

### About the testng.xml and parallelization
Yes! If you change the `value` attribute to `chrome` it will start your tests using Google Chrome.

You can run parallel tests as explained here: https://github.com/eliasnogueira/selenium-java-lean-test-architecture?tab=readme-ov-file#parallel-execution

You can use this sample `testng.xml` that will run the tests in parallel for these 3 different browsers. Remember that you need to read the parallel execution approach mentioned in the link above.

```xml
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="Multi-browser test" verbose="1" parallel="tests" thread-count="6">

    <test name="Chrome execution">
        <parameter name="browser" value="chrome"/>
        <classes>
            <class name="com.eliasnogueira.test.BookRoomWebTest"/>
        </classes>
    </test>

    <test name="Firefox Execution">
        <parameter name="browser" value="firefox"/>
        <classes>
            <class name="com.eliasnogueira.test.BookRoomWebTest"/>
        </classes>
    </test>

    <test name="Edge Execution">
        <parameter name="browser" value="edge"/>
        <classes>
            <class name="com.eliasnogueira.test.BookRoomWebTest"/>
        </classes>
    </test>
</suite>
```

I hope that it helps!

**Reply by DamianSzp** [reply]
Hi @eliasnogueira 
Could you share some examples of how the Explicitly Wait can be implemented or some thoughts what is the best way to do it?

**Reply by eliasnogueira** [reply]
Hi @DamianSzp 
Could you open a new discussion topic to not to mix the subject?

Thank you!
